File: java/scon-core/src/main/java/io/github/realmlabs/scon/SconMapper.java
package io.github.realmlabs.scon;

import java.lang.reflect.*;
import java.nio.file.Path;
import java.util.*;

public final class SconMapper {
    private SconMapper() {}

    public static <T> T readValue(String source, Class<T> type) {
        return decode(Scon.parseString(source), type);
    }

    public static <T> T readFile(Path path, Class<T> type) {
        return decode(Scon.parseFile(path), type);
    }

    public static String writeValue(Object value) {
        return Scon.formatValue(encode(value));
    }

    @SuppressWarnings("unchecked")
    public static <T> T decode(SconValue value, Type type) {
        if (type == Object.class) return (T) plain(value);
        if (type == String.class) {
            if (value instanceof SconString s) return (T) s.value();
            throw serde("expected string");
        }
        if (type == boolean.class || type == Boolean.class) {
            if (value instanceof SconBool b) return (T) Boolean.valueOf(b.value());
            throw serde("expected bool");
        }
        if (type instanceof Class<?> cls && Number.class.isAssignableFrom(box(cls)) || isPrimitiveNumber(type)) {
            if (!(value instanceof SconNumber number)) throw serde("expected number");
            return (T) decodeNumber(number, (Class<?>) type);
        }
        if (type instanceof Class<?> cls && cls.isEnum()) {
            if (!(value instanceof SconString s)) throw serde("expected enum string");
            return (T) Enum.valueOf((Class<? extends Enum>) cls.asSubclass(Enum.class), s.value());
        }
        if (type instanceof Class<?> cls && cls.isRecord()) return decodeRecord(value, cls);
        if (type instanceof ParameterizedType pt && pt.getRawType() instanceof Class<?> raw) {
            if (List.class.isAssignableFrom(raw)) {
                if (!(value instanceof SconArray array)) throw serde("expected array");
                var out = new ArrayList<>();
                Type itemType = pt.getActualTypeArguments()[0];
                for (SconValue item : array) out.add(decode(item, itemType));
                return (T) out;
            }
            if (Map.class.isAssignableFrom(raw)) {
                if (!(value instanceof SconObject object)) throw serde("expected object");
                if (pt.getActualTypeArguments()[0] != String.class) throw serde("SCON map keys must be strings");
                var out = new LinkedHashMap<String, Object>();
                Type itemType = pt.getActualTypeArguments()[1];
                for (var entry : object.entrySet()) out.put(entry.getKey(), decode(entry.getValue(), itemType));
                return (T) out;
            }
        }
        if (type instanceof Class<?> cls && cls.isArray()) {
            if (!(value instanceof SconArray array)) throw serde("expected array");
            Class<?> itemType = cls.getComponentType();
            Object out = Array.newInstance(itemType, array.size());
            for (int i = 0; i < array.size(); i++) Array.set(out, i, decode(array.get(i), itemType));
            return (T) out;
        }
        if (type instanceof Class<?> cls) return decodeBean(value, cls);
        throw serde("unsupported target type: " + type);
    }

    public static SconValue encode(Object value) {
        if (value == null) return SconNull.INSTANCE;
        if (value instanceof SconValue sconValue) return sconValue;
        if (value instanceof Boolean bool) return new SconBool(bool);
        if (value instanceof String string) return new SconString(string);
        if (value instanceof Byte || value instanceof Short || value instanceof Integer || value instanceof Long) {
            long n = ((Number) value).longValue();
            return n < 0 ? SconNumber.ofI64(n) : SconNumber.ofU64(n);
        }
        if (value instanceof Float || value instanceof Double) {
            double n = ((Number) value).doubleValue();
            if (!Double.isFinite(n)) throw serde("non-finite floats cannot be serialized");
            return SconNumber.ofF64(n);
        }
        if (value instanceof Enum<?> e) return new SconString(e.name());
        if (value instanceof Iterable<?> iterable) {
            var out = new SconArray();
            for (Object item : iterable) out.add(encode(item));
            return out;
        }
        if (value.getClass().isArray()) {
            var out = new SconArray();
            int len = Array.getLength(value);
            for (int i = 0; i < len; i++) out.add(encode(Array.get(value, i)));
            return out;
        }
        if (value instanceof Map<?, ?> map) {
            var out = new SconObject();
            for (var entry : map.entrySet()) {
                if (!(entry.getKey() instanceof String key)) throw serde("SCON map keys must be strings");
                out.put(key, encode(entry.getValue()));
            }
            return out;
        }
        if (value.getClass().isRecord()) return encodeRecord(value);
        return encodeBean(value);
    }

    private static Object decodeNumber(SconNumber number, Class<?> type) {
        if (type == byte.class || type == Byte.class) {
            long n = number.asI64();
            if (n < Byte.MIN_VALUE || n > Byte.MAX_VALUE) throw serde("integer overflow");
            return (byte) n;
        }
        if (type == short.class || type == Short.class) {
            long n = number.asI64();
            if (n < Short.MIN_VALUE || n > Short.MAX_VALUE) throw serde("integer overflow");
            return (short) n;
        }
        if (type == int.class || type == Integer.class) {
            long n = number.asI64();
            if (n < Integer.MIN_VALUE || n > Integer.MAX_VALUE) throw serde("integer overflow");
            return (int) n;
        }
        if (type == long.class || type == Long.class) return number.asI64();
        if (type == float.class || type == Float.class) return (float) number.asF64();
        if (type == double.class || type == Double.class) return number.asF64();
        throw serde("unsupported numeric type");
    }

    private static <T> T decodeRecord(SconValue value, Class<?> cls) {
        if (!(value instanceof SconObject object)) throw serde("expected object");
        try {
            var components = cls.getRecordComponents();
            var args = new Object[components.length];
            var types = new Class<?>[components.length];
            for (int i = 0; i < components.length; i++) {
                var component = components[i];
                SconValue item = object.get(component.getName());
                if (item == null) throw serde("missing field " + component.getName());
                args[i] = decode(item, component.getGenericType());
                types[i] = component.getType();
            }
            var constructor = cls.getDeclaredConstructor(types);
            constructor.setAccessible(true);
            @SuppressWarnings("unchecked")
            T out = (T) constructor.newInstance(args);
            return out;
        } catch (ReflectiveOperationException ex) {
            throw serde("record decode failed: " + ex.getMessage());
        }
    }

    private static <T> T decodeBean(SconValue value, Class<?> cls) {
        if (!(value instanceof SconObject object)) throw serde("expected object");
        try {
            Object out = cls.getDeclaredConstructor().newInstance();
            for (Field field : cls.getDeclaredFields()) {
                if (Modifier.isStatic(field.getModifiers())) continue;
                SconValue item = object.get(field.getName());
                if (item == null) continue;
                field.setAccessible(true);
                field.set(out, decode(item, field.getGenericType()));
            }
            @SuppressWarnings("unchecked")
            T typed = (T) out;
            return typed;
        } catch (ReflectiveOperationException ex) {
            throw serde("bean decode failed: " + ex.getMessage());
        }
    }

    private static SconValue encodeRecord(Object value) {
        var out = new SconObject();
        try {
            for (RecordComponent component : value.getClass().getRecordComponents()) {
                var accessor = component.getAccessor();
                accessor.setAccessible(true);
                out.put(component.getName(), encode(accessor.invoke(value)));
            }
            return out;
        } catch (ReflectiveOperationException ex) {
            throw serde("record encode failed: " + ex.getMessage());
        }
    }

    private static SconValue encodeBean(Object value) {
        var out = new SconObject();
        try {
            for (Field field : value.getClass().getDeclaredFields()) {
                if (Modifier.isStatic(field.getModifiers())) continue;
                field.setAccessible(true);
                out.put(field.getName(), encode(field.get(value)));
            }
            return out;
        } catch (ReflectiveOperationException ex) {
            throw serde("bean encode failed: " + ex.getMessage());
        }
    }

    private static Object plain(SconValue value) {
        if (value == SconNull.INSTANCE) return null;
        if (value instanceof SconBool bool) return bool.value();
        if (value instanceof SconString string) return string.value();
        if (value instanceof SconNumber number) return number.plainValue();
        if (value instanceof SconArray array) {
            var out = new ArrayList<>();
            for (SconValue item : array) out.add(plain(item));
            return out;
        }
        if (value instanceof SconObject object) {
            var out = new LinkedHashMap<String, Object>();
            for (var entry : object.entrySet()) out.put(entry.getKey(), plain(entry.getValue()));
            return out;
        }
        throw serde("unsupported SCON value");
    }

    private static boolean isPrimitiveNumber(Type type) {
        return type == byte.class || type == short.class || type == int.class || type == long.class || type == float.class || type == double.class;
    }

    private static Class<?> box(Class<?> cls) {
        if (!cls.isPrimitive()) return cls;
        if (cls == int.class) return Integer.class;
        if (cls == long.class) return Long.class;
        if (cls == short.class) return Short.class;
        if (cls == byte.class) return Byte.class;
        if (cls == float.class) return Float.class;
        if (cls == double.class) return Double.class;
        if (cls == boolean.class) return Boolean.class;
        return cls;
    }

    private static SconException serde(String message) {
        return new SconException(ErrorCode.Serde, message);
    }
}

File: csharp/src/Scon.Core/SconMapper.cs
namespace RealmLabs.Scon;

public static class SconMapper
{
    public static T Deserialize<T>(string source) => (T)Decode(Scon.ParseString(source), typeof(T))!;
    public static string Serialize<T>(T value) => Scon.FormatValue(Encode(value));
    public static object? ToPlain(SconValue value) => value switch { SconNull => null, SconBool b => b.Value, SconString s => s.Value, SconNumberValue n => n.Number.PlainValue(), SconArray a => a.Select(ToPlain).ToList(), SconObject o => o.ToDictionary(kv => kv.Key, kv => ToPlain(kv.Value)), _ => throw new SconException(ErrorCode.Serde, "unsupported SCON value") };
    public static SconValue Encode(object? value)
    {
        if (value is null) return SconNull.Instance;
        if (value is SconValue scon) return scon;
        if (value is JsonElement json) return EncodeJsonElement(json);
        if (value is bool b) return new SconBool(b);
        if (value is string s) return new SconString(s);
        if (value is byte or sbyte or short or ushort or int or uint or long) { var n = Convert.ToInt64(value, CultureInfo.InvariantCulture); return new SconNumberValue(n < 0 ? SconNumber.FromI64(n) : SconNumber.FromU64((ulong)n)); }
        if (value is ulong ul) return new SconNumberValue(SconNumber.FromU64(ul));
        if (value is float or double or decimal) { var d = Convert.ToDouble(value, CultureInfo.InvariantCulture); if (!double.IsFinite(d)) throw new SconException(ErrorCode.Serde, "non-finite floats cannot be serialized"); return new SconNumberValue(SconNumber.FromF64(d)); }
        if (value is Enum e) return new SconString(e.ToString());
        if (value is IDictionary dict) { var obj = new SconObject(); foreach (DictionaryEntry entry in dict) { if (entry.Key is not string key) throw new SconException(ErrorCode.Serde, "SCON map keys must be strings"); obj.Set(key, Encode(entry.Value)); } return obj; }
        if (value is IEnumerable enumerable && value is not string) return new SconArray(enumerable.Cast<object?>().Select(Encode));
        var outObj = new SconObject();
        foreach (var prop in value.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public).Where(p => p.GetIndexParameters().Length == 0 && p.CanRead)) outObj.Set(prop.Name, Encode(prop.GetValue(value)));
        foreach (var field in value.GetType().GetFields(BindingFlags.Instance | BindingFlags.Public)) outObj.Set(field.Name, Encode(field.GetValue(value)));
        return outObj;
    }
    public static object? Decode(SconValue value, Type type)
    {
        var underlying = Nullable.GetUnderlyingType(type);
        if (underlying is not null) return value is SconNull ? null : Decode(value, underlying);
        if (type == typeof(object)) return ToPlain(value);
        if (type == typeof(string)) return value is SconString s ? s.Value : throw new SconException(ErrorCode.Serde, "expected string");
        if (type == typeof(bool)) return value is SconBool b ? b.Value : throw new SconException(ErrorCode.Serde, "expected bool");
        if (type.IsEnum) return value is SconString s ? Enum.Parse(type, s.Value) : throw new SconException(ErrorCode.Serde, "expected enum string");
        if (IsNumberType(type)) return DecodeNumber(value, type);
        if (type.IsArray) { if (value is not SconArray a) throw new SconException(ErrorCode.Serde, "expected array"); var itemType = type.GetElementType()!; var outArray = Array.CreateInstance(itemType, a.Count); for (var i = 0; i < a.Count; i++) outArray.SetValue(Decode(a[i], itemType), i); return outArray; }
        if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>)) { if (value is not SconArray a) throw new SconException(ErrorCode.Serde, "expected array"); var itemType = type.GetGenericArguments()[0]; var list = (IList)Activator.CreateInstance(type)!; foreach (var item in a) list.Add(Decode(item, itemType)); return list; }
        if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Dictionary<,>)) { if (value is not SconObject o) throw new SconException(ErrorCode.Serde, "expected object"); var args = type.GetGenericArguments(); if (args[0] != typeof(string)) throw new SconException(ErrorCode.Serde, "SCON map keys must be strings"); var dict = (IDictionary)Activator.CreateInstance(type)!; foreach (var (key, item) in o) dict.Add(key, Decode(item, args[1])); return dict; }
        if (value is not SconObject obj) throw new SconException(ErrorCode.Serde, "expected object");
        var ctor = type.GetConstructors().OrderByDescending(c => c.GetParameters().Length).FirstOrDefault();
        object instance;
        if (ctor is not null && ctor.GetParameters().Length > 0)
        {
            var args = ctor.GetParameters().Select(p => obj.Get(p.Name!) is { } item ? Decode(item, p.ParameterType) : throw new SconException(ErrorCode.Serde, $"missing field {p.Name}")).ToArray();
            instance = ctor.Invoke(args);
        }
        else instance = Activator.CreateInstance(type) ?? throw new SconException(ErrorCode.Serde, $"unsupported target type {type}");
        foreach (var prop in type.GetProperties(BindingFlags.Instance | BindingFlags.Public).Where(p => p.CanWrite)) if (obj.Get(prop.Name) is { } item) prop.SetValue(instance, Decode(item, prop.PropertyType));
        foreach (var field in type.GetFields(BindingFlags.Instance | BindingFlags.Public)) if (obj.Get(field.Name) is { } item) field.SetValue(instance, Decode(item, field.FieldType));
        return instance;
    }
    private static object DecodeNumber(SconValue value, Type type)
    {
        if (value is not SconNumberValue number) throw new SconException(ErrorCode.Serde, "expected number");
        checked { if (type == typeof(byte)) return (byte)number.Number.AsU64(); if (type == typeof(sbyte)) return (sbyte)number.Number.AsI64(); if (type == typeof(short)) return (short)number.Number.AsI64(); if (type == typeof(ushort)) return (ushort)number.Number.AsU64(); if (type == typeof(int)) return (int)number.Number.AsI64(); if (type == typeof(uint)) return (uint)number.Number.AsU64(); if (type == typeof(long)) return number.Number.AsI64(); if (type == typeof(ulong)) return number.Number.AsU64(); }
        if (type == typeof(float)) return (float)number.Number.AsF64();
        if (type == typeof(double)) return number.Number.AsF64();
        throw new SconException(ErrorCode.Serde, "unsupported numeric type");
    }
    private static bool IsNumberType(Type type) => type == typeof(byte) || type == typeof(sbyte) || type == typeof(short) || type == typeof(ushort) || type == typeof(int) || type == typeof(uint) || type == typeof(long) || type == typeof(ulong) || type == typeof(float) || type == typeof(double);
    private static SconValue EncodeJsonElement(JsonElement value) => value.ValueKind switch
    {
        JsonValueKind.Null => SconNull.Instance,
        JsonValueKind.True => new SconBool(true),
        JsonValueKind.False => new SconBool(false),
        JsonValueKind.String => new SconString(value.GetString()!),
        JsonValueKind.Number when value.TryGetInt64(out var i64) => new SconNumberValue(i64 < 0 ? SconNumber.FromI64(i64) : SconNumber.FromU64((ulong)i64)),
        JsonValueKind.Number when value.TryGetUInt64(out var u64) => new SconNumberValue(SconNumber.FromU64(u64)),
        JsonValueKind.Number => new SconNumberValue(SconNumber.FromF64(value.GetDouble())),
        JsonValueKind.Array => new SconArray(value.EnumerateArray().Select(EncodeJsonElement)),
        JsonValueKind.Object => EncodeJsonObject(value),
        _ => throw new SconException(ErrorCode.Serde, "unsupported JSON value"),
    };
    private static SconObject EncodeJsonObject(JsonElement value)
    {
        var obj = new SconObject();
        foreach (var prop in value.EnumerateObject()) obj.Set(prop.Name, EncodeJsonElement(prop.Value));
        return obj;
    }
}
